### ISSUE TITLE
Tail call optimization does not seem to work?

### ISSUE THREAD

**Issue by tomaskala** [Robert-van-Engelen/tinylisp#15]
Hello, I've been experimenting with the tail call-optimized version of tinylisp (`tinylisp-opt.c`), but it doesn't seem to optimize tail calls. Specifically, I've been testing it with a tail call-optimized factorial computation:
```lisp
(define factorial1
  (lambda (n acc)
    (if (< n 2) acc
      (factorial1 (- n 1) (* n acc)))))

(define factorial
  (lambda (n)
    (factorial1 n 1)))
```

The following shows a session of me compiling `tinylisp-opt.c` and running the above `factorial` function with the value of 1000:
```bash
$ gcc -o tinylisp tinylisp-opt.c
$ ./tinylisp
tinylisp
930>(define factorial1
(lambda (n acc)
(if (< n 2) acc
(factorial1 (- n 1) (* n acc)))))
factorial1
871>(define factorial
(lambda (n)
(factorial1 n 1)))
factorial
842>(factorial 1000)
Aborted (core dumped)
```

I'd have expected the computation to finish successfully and return `inf`, since the result exceeds the valid `double` range.

When I compare it with your [full LISP implementation](https://github.com/Robert-van-Engelen/lisp/) which is also tail call-optimized, I'm getting the correct results:
```bash
$ gcc -o lisp lisp.c
$ ./lisp
lisp
8020+2000>(define factorial1
?(lambda (n acc)
?(if (< n 2) acc
?(factorial1 (- n 1) (* n acc)))))
factorial1
7976+1996>(define factorial
?(lambda (n)
?(factorial1 n 1)))
factorial
7960+1994>(factorial 1000)
inf
```

I've even tried to increase the available memory of `tinylisp` by setting `N` to larger values, but there is always a value that makes the code run out of memory. Compared with the full LISP implementation which happily calculates `(factorial 100000)` (though again producing `inf`, of course), this makes me think that there is something missing from `tinylisp-opt.c` to make it fully tail call-optimized.

**Comment by Robert-van-Engelen** [comment]
I think you're right and it is because I never bothered to update tinylisp after writing the other lisp ([lisp in 1K lines of C](https://github.com/Robert-van-Engelen/lisp)). Those small lisp interpreters do perform full TC optimization:
~~~
6262+1924>(factorial 1000000)
inf
6262+1924>
~~~
Tinylisp does a bit of TC, but not full TC. What is missing is a TC classification of lisp primitives to TC-optimize them (`let`, `cond`, `if`, `begin` etc). So the `if` will _not properly tail-call_ in your `factorial1`. Note that the other lisp I wrote include `TAILCALL` designations for several primitives, defined in the function primitives `prim[]` table. Those return an unevaluated lisp expression that the eval loop then continues to operate on (to avoid a recursive eval call).

We can (should) do the same for tinylisp.

**Comment by tomaskala** [comment]
The TC classification seems to be in place, though: There is a `short t` flag for each primitive with `1` denoting tail calls in the [prim array](https://github.com/Robert-van-Engelen/tinylisp/blob/main/src/tinylisp-opt.c#L56), and a conditional to keep evaluating such primitives in the [evaluation loop](https://github.com/Robert-van-Engelen/tinylisp/blob/main/src/tinylisp-opt.c#L69).

**Comment by Robert-van-Engelen** [comment]
Yes, but it doesn't work the same way and so may need the same eval loop update as the "big brother" lisp.

**Comment by Robert-van-Engelen** [comment]
I've added a clarification of TC optimization limitations to the PDF page 35 related to argument evaluations that will accumulate in memory.

I'm thinking that there should be a way fix "ABC" garbage collection to make it work. But it is not just sufficient for data structures to be acyclic as the SectorLisp author presents it. Acyclic forms like `(lambda (x . args) args)` (essentially `cdr`) do not work with "ABC", because the evaluated `args` is removed from the stack when the closure returns (or tail-calls).

**Comment by tomaskala** [comment]
Thank you for the clarification! I was actually following along your PDF with my own implementation, and was scratching my head why tail calls do not get optimized. I had made a couple of customizations, but hadn't yet gotten to implementing a full garbage collection, so I wasn't sure if I made a mistake in the evaluator implementation or not. Good to know the underlying reason, thanks a lot!

**Comment by karme** [comment]
the test above was against the wrong version. But using the correct version it doesn't work either:

```
 $ git diff tinylisp-float-opt.c
diff --git a/src/tinylisp-float-opt.c b/src/tinylisp-float-opt.c
index 5480686..6a0f254 100644
--- a/src/tinylisp-float-opt.c
+++ b/src/tinylisp-float-opt.c
@@ -2,6 +2,7 @@
 #include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
+#include <stdint.h>
 #define I unsigned
 #define L float
 #define T(x) *(uint32_t*)&x>>20
karme@amalthea:~/develop/tinylisp/src$ { cat <<EOF
(define even? (lambda(n) (if (eq? n 0) #t (odd? (- n 1)))))
(define odd? (lambda(n) (if (eq? n 0) #f (even? (- n 1)))))
(even? 2)
(odd? 3)
(even? 100000)
EOF
  }|./tinylisp-float-opt
tinylisp
913>even?
864>odd?
817>#t
817>#t
Aborted
```

**Comment by Robert-van-Engelen** [comment]
Good question. But please note that the TR check is limited to self-recursive lambda calls only. Mutual TR is not handled in tinylisp. It is not a full compiler with optimizer.
